### PR TITLE
upgrade for rainlang rust

### PR DESCRIPTION
**Pull request by rouzwelt**
port latest from `rainprotocol/meta` repo at `master` branch

### CODE REVIEW

**Review by thedavidmeister — CHANGES_REQUESTED** [review]
generally need waaaaay more tests

File: cli/src/lib.rs [hunk 1]
```diff
@@ -1,3 +1,4 @@
-pub mod cli;
+// pub mod cli;
```
[the remark this answers is not in the document]
Comment (rouzwelt) [reply cli/src/lib.rs:1]: reason being it offers no functionality for the lib.rs, it is basicalll just for binary

[the remark this answers is not in the document]
Comment (rouzwelt) [reply cli/src/lib.rs:1]: I wasn't sure if we should remove it or not, so commented it to ask you, if you think we should have it, feel free to uncomment it

File: cli/test/abis/valid/IInterpreterV1.json [hunk 2]
```diff
@@ -0,0 +1,63 @@
+{
```
Comment (thedavidmeister) [inline cli/test/abis/valid/IInterpreterV1.json:1]: this is not an abi, its a compilation artifact that _includes_ the abi as one of the keys

also, is there no way to have these built locally and referenced rather than committing many thousands of lines of code to the repo?

File: cli/src/utils/mod.rs [hunk 3]
```diff
@@ -0,0 +1,75 @@
+/// converts string to bytes32
```
Comment (thedavidmeister) [inline cli/src/utils/mod.rs:1]: don't have `utils` folder with undifferentiated code

File: cli/src/meta/types/solidity_abi/v2.rs [hunk 4]
```diff
@@ -1,30 +1,94 @@
...
 
-impl Validate for SolidityAbi {
+impl Validate for SolidityAbiMeta {
     fn validate(&self) -> Result<(), ValidationErrors> {
-        ValidationErrors::merge_all(
-            Ok(()),
-            "root",
-            self.0.iter().map(|item| item.validate()).collect()
-        )
+        // ValidationErrors::merge_all(
```
Comment (thedavidmeister) [inline cli/src/meta/types/solidity_abi/v2.rs:32]: commented out code

File: cli/src/meta/types/rainlang/v1.rs [hunk 5]
```diff
@@ -0,0 +1,2 @@
+/// # Rainlang Meta
+pub type RainlangMeta = String;
```
Comment (thedavidmeister) [inline cli/src/meta/types/rainlang/v1.rs:2]: files missing final newline char

File: cli/src/cli/mod.rs [hunk 6]
```diff
@@ -26,19 +26,22 @@ pub enum Meta {
...
     Solc(solc::Solc),
+    #[command(subcommand)]
+    Subgraph(subgraph::Sg)
 }
 
-pub fn dispatch(meta: Meta) -> Result<()> {
+pub fn dispatch(meta: Meta) -> anyhow::Result<()> {
     match meta {
-        Meta::Schema(schema) => schema::dispatch(schema),
+        Meta::Build(build)          => build::build(build),
```
Comment (thedavidmeister) [inline cli/src/cli/mod.rs:35]: weird formatting, did you run `cargo fmt` over this?

File: cli/src/cli/build.rs [hunk 7]
```diff
@@ -84,20 +87,20 @@ impl TryFrom<&BuildItem> for RainMetaDocumentV1Item {
     }
 }
 
-impl BuildItem {
-    /// Write a BuildItem to a byte buffer as normalized, encoded cbor rain meta.
-    fn write<W: std::io::Write>(&self, writer: &mut W) -> anyhow::Result<()> {
-        Ok(ciborium::into_writer(&RainMetaDocumentV1Item::try_from(self)?, writer)?)
-    }
-}
+// impl BuildItem {
```
Comment (thedavidmeister) [inline cli/src/cli/build.rs:90]: commented code

File: cli/src/meta/mod.rs [hunk 8]
```diff
@@ -52,32 +69,63 @@ pub enum ContentEncoding {
...
     None,
     En,
 }
 
-#[derive(PartialEq, Debug)]
-pub struct RainMetaDocumentV1Item {
+/// # Rain Meta Map
+/// represents a rain meta data and configuration that can be cbor encoded or unpacked back to the meta types
+#[derive(PartialEq, Debug, Clone)]
+pub struct MetaMap {
```
Comment (thedavidmeister) [inline cli/src/meta/mod.rs:103]: why was this renamed? "meta map" means nothing to me, the original name was descriptive and versioned

if we have a v2 of rain document that also has maps, this "meta map" name will cause problems

**Review by thedavidmeister — CHANGES_REQUESTED** [review]

File: cli/src/utils/mod.rs (hunk 3 above)
Comment (rouzwelt) [reply cli/src/utils/mod.rs:1]: well this is a util, there is not certain place for it as far as I see it, can be used anywhere for converting a bytes32 to str and back,

File: cli/src/meta/mod.rs (hunk 8 above)
Comment (rouzwelt) [reply cli/src/meta/mod.rs:134]: well the reason I changed it because RaimMetaDocument explicitly refers to a meta seq with Raindocument magic number, but this is just a single meta map, but feel free to update it as you see fit

File: cli/src/utils/mod.rs (hunk 3 above)
Comment (thedavidmeister) [reply cli/src/utils/mod.rs:1]: @rouzwelt so name it as such, make a module for converting bytes32

File: cli/src/meta/mod.rs (hunk 8 above)
Comment (thedavidmeister) [reply cli/src/meta/mod.rs:134]: @rouzwelt but the original struct was rain meta document _item_, as in a single item in the seq

File: cli/test/abis/valid/IInterpreterV1.json (hunk 2 above)
Comment (thedavidmeister) [reply cli/test/abis/valid/IInterpreterV1.json:1]: @rouzwelt not committing them to the repo, is there a way they can be automatically built locally, e.g. with a `build.rs` file or similar?

Comment (rouzwelt) [reply cli/test/abis/valid/IInterpreterV1.json:1]: hmm, there should be a way ofc, not sure where to source those data from though, maybe we can use the built artifacts of the sol files within the same repo

File: cli/src/meta/mod.rs (hunk 8 above)
Comment (rouzwelt) [reply cli/src/meta/mod.rs:134]: yea so having that prefix + "item" sounds a bit vague to me, but it's not deal breaker, I just personally thought this name sounds better, but if you think the former was better, let me know

Comment (thedavidmeister) [reply cli/src/meta/mod.rs:134]: @rouzwelt i'd like to use the original name, or something equally specific and versioned
